lib/Modules/HyperVCIHooks/HyperVCIHooks.psm1

```diff
@@ -0,0 +1,1538 @@
+
+    if (!(Get-WindowsFeature hyper-v).Installed) {
+        Install-WindowsFeature -Name Hyper-V -includemanagementtools
+        Invoke-JujuReboot -Now
+    }
+
+    # Disable firewall
+    Start-ExternalCommand { netsh.exe advfirewall set allprofiles state off } -ErrorMessage "Failed to disable firewall."
+    # Disable automatic updates
+    Write-JujuLog "Disabling automatic updates"
+    $update_service = Get-WmiObject Win32_Service -Filter 'Name="wuauserv"'
+    $update_service.ChangeStartMode("Disabled")
+    $update_service.StopService()
+
+    # Allow self-signed drivers installation
+    $bcdedit = "C:\bcdedit.txt"
+    $signingStatus = Get-JujuCharmConfig -Scope "test-signing"
+    if ((!(Test-Path -path $bcdedit)) -and ($signingStatus)){
+        bcdedit -set loadoptions DDISABLE_INTEGRITY_CHECKS
+        bcdedit -set TESTSIGNING ON
+        New-Item -ItemType file $bcdedit
+        Invoke-JujuReboot
+    }
+    
+    # Set Administrator user password
+    $admin_user = "Administrator"
+    $admin_password = Get-JujuCharmConfig -Scope 'administrator-password'
+    net user $admin_user "$admin_password"
+    
+    Write-JujuLog "Enable and start MSiSCSI"
+    $msiscsi_service = Get-WmiObject Win32_Service -Filter 'Name="MSiSCSI"'
+    $msiscsi_service.ChangeStartMode("Automatic")
+    $msiscsi_service.StartService()
+
+    Import-CloudbaseCert
+    Start-ConfigureVMSwitch
+    Write-PipConfigFile
+
+    # Install Git
+    Install-Dependency 'git-url' @('/SILENT')
+    Add-ToUserPath "${env:ProgramFiles(x86)}\Git\cmd"
+    Add-ToSystemPath "${env:ProgramFiles(x86)}\Git\cmd"
+
+    # Install Python 2.7.x (x86)
+    Install-Dependency 'python27-url' @('/qn')
+    Add-ToUserPath "${env:SystemDrive}\Python27;${env:SystemDrive}\Python27\scripts"
+    Add-ToSystemPath "${env:SystemDrive}\Python27;${env:SystemDrive}\Python27\scripts"
+
+    # Install FreeRDP Hyper-V console access
+    $enableFreeRDP = Get-JujuCharmConfig -Scope 'enable-freerdp-console'
+    if ($enableFreeRDP -eq $true) {
+        Install-FreeRDPConsole
+    }
+
+    Write-JujuLog "Installing pip"
+    $conf_pip_version = Get-JujuCharmConfig -Scope 'pip-version'
+    $tmpPath = Join-Path $env:TEMP "get-pip.py"
+    Start-DownloadFile -Uri "https://bootstrap.pypa.io/get-pip.py" -SkipIntegrityCheck -OutFile $tmpPath
+    Start-ExternalCommand -ScriptBlock { python $tmpPath $conf_pip_version } -ErrorMessage "Failed to install pip."
+    Remove-Item $tmpPath
+    $version = Start-ExternalCommand { pip.exe --version } -ErrorMessage "Failed to get pip version."
+    Write-JujuLog "Pip version: $version"
+
+    Write-JujuLog "Installing pip dependencies"
+    $pythonPkgs = Get-JujuCharmConfig -Scope 'extra-python-packages'
+    if ($pythonPkgs) {
+        $pythonPkgsArr = $pythonPkgs.Split()
+        foreach ($pythonPkg in $pythonPkgsArr) {
+            Write-JujuLog "Installing $pythonPkg"
+            Start-ExternalCommand -ScriptBlock { pip install -U $pythonPkg } `
```

Comment:
Updated to first generate a requirments.txt file from all the extra-packages and then run only one exec call